### DISCUSSION TITLE
Towny Integration (feature request)

### DISCUSSION THREAD

**Discussion by roracle** [QuickShop-Community/QuickShop-Hikari#372] [Ideas]
Integration with Towny fake bank accounts. Often used to hold money and purchase land and other Towny-related things, a growing town could benefit from selling items and increasing its own value in the game.

1. Place a chest
2. While looking at the chest, type `/qs town` (or `/qs nation`)
3. QS checks if the person running the command is in a town and if so are they the mayor? (or king if nation)
4. QS creates a shop only the mayor (or king) can manipulate.
5. Shops use a list of supported items from the config file or any if bankmode is disabled.

The sign shop should label the town as the buyer/seller.

Before we move on to the config options, it became more apparent how maybe this would be better as an addon for QS instead of making it part of it ultimately.  If anything, it could be built in and followed up with a separate config file.  It has utility, and my ideas give it room to be utilized in a more specific way as well.  Anyways, here are config options:

```
# IMPORTANT: Go to Towny Permissions Config
# and add quicksshop.town to mayor permissions
# and add quicksshop.nation to king permissions.
# Failure to do this will render this feature useless.
# UNLESS QS has a way around this, of course!
# (like checking if the person using the command
# is the town mayor or nation king, and if that can
# be done, this warning can be ignored and the
# permissions can be waived.)

# Towny Integration, if enabled set to true.
use-towny: true

# the settings for these should be exactly
# as they are in your Towny config. If QS can
# determine this on its own, this setting could
# be eliminated.
town-account-prefix: town-
nation-account-prefix: nation-

# Should QS taxes go to the town the shop
# is in?  If shop isn't in a town, taxes go
# to the server account.
taxes-to-town: true

# Should a town or nation owned quickshop only be
# created in a bank plot owned by the town
# or nation? Towny settings may allow otherwise.
respect-bank-plot: true

# ****************************************
# Above is all the config options for basic
# usage.
# ****************************************
```

Next could be a separate config file to make this easier to read.

```
# ****************************************
# Below  will force towns into being used as
# banking systems where they exchange rare,
# non-renewable items for money credits.
# ****************************************

# Bankmode forces towns or nations to be used as
# banks if Towny integration is being used. This means
# they can only buy or sell things on the list below.
# Everything below this will not be affected if this is false.
# If use-towny: false then this will not work.
bankmode: true

# this will blacklist the below list from
# being sold by players.
blacklist-items: true

# set to true if you want the bank item prices forced
# (to regulate town and nation economic activity)
# setting to false will allow mayors/kings to set
# prices themselves.  True is recommended for a 
# stable economic system.
force-prices: true

# sell increase percentage. This will sell the item
# in the below list at a higher rate.  Values 0 through 100
# Example being if Emerald ore is worth $1000 to the bank,
# it will sell it at 10% higher ($1100) so the bank can make money.
sell-percent: 10.0

# Below is the bankmode item and price list.
# Items on the QS blacklist cannot be used here.
# Items set to a town or nation shop automatically
# get priced at this value for buying shops.
# Edit to your liking. Default settings are based on
# rarity, chance of drops from all levels of fortune,
# and usefulness as a whole.
COAL_ORE: 0.01
COPPER_ORE: 0.05
IRON_ORE: 0.10
LAPIS_ORE: 0.25
REDSTONE_ORE: 1.00
GOLD_ORE: 5.00
DIAMOND_ORE: 50.00
EMERALD_ORE: 1000.00
DEEPSLATE_COAL_ORE: 0.01
DEEPSLATE_COPPER_ORE: 0.05
DEEPSLATE_IRON_ORE: 0.10
DEEPSLATE_LAPIS_ORE: 0.25
DEEPSLATE_REDSTONE_ORE: 1.00
DEEPSLATE_GOLD_ORE: 5.00
DEEPSLATE_DIAMOND_ORE: 50.00
DEEPSLATE_EMERALD_ORE: 1000.00
# NETHER_QUARTZ_ORE: 0.50
# NETHER_GOLD_ORE: 1.25
ANCIENT_DEBRIS: 500.00
# HEART_OF_THE_SEA: 25000.00
# ENCHANTED_GOLDEN_APPLE: 250000.00

```

**Extra information, slightly related:**
Ores are used by default because they are unable to be farmed.
Players are encouraged to build their town's bank account as much as possible to use this feature.
The best option is to have player payments (like Jobs payments) go to the town accounts (all or taxed).

**Comment by Ghost-chu** [comment]
Track the progress here: https://github.com/Ghost-chu/QuickShop-Hikari/issues/374

**Comment by Ghost-chu** [comment]
Does commands `/qs town` and `/qs nation` will make shops to a BankShop and apply the BankMode rules in your idea?

**Reply by roracle** [reply]
Yes, if bankmode is active. If it's not active it's treated like a regular shop but with the town or nation accounts, changeable by the mayor or king.

**Comment by roracle** [comment]
I re-evaluated the rarity and possible cost of ores. I forgot to keep in mind the deepslate variants in my calculations.
Copper, because it has a chance to drop up to 20 raw copper, makes it the least valuable.  Iron ingots can be farmed easier than coal, so iron is worth less than coal.  This makes sense to have coal cost more since it's such a useful fuel type. Diamond Ore, Emerald Ore, Ancient Debris, Heart of the Sea, and Enchanted Golden Apple have been adjusted accordingly as well.
```
COPPER_ORE: 0.01
IRON_ORE: 0.05
COAL_ORE: 0.10
LAPIS_ORE: 0.25
REDSTONE_ORE: 1.00
GOLD_ORE: 5.00
DIAMOND_ORE: 20.00
EMERALD_ORE: 500.00
DEEPSLATE_COPPER_ORE: 0.01
DEEPSLATE_IRON_ORE: 0.05
DEEPSLATE_COAL_ORE: 0.10
DEEPSLATE_LAPIS_ORE: 0.25
DEEPSLATE_REDSTONE_ORE: 1.00
DEEPSLATE_GOLD_ORE: 5.00
DEEPSLATE_DIAMOND_ORE: 20.00
DEEPSLATE_EMERALD_ORE: 500.00
# NETHER_QUARTZ_ORE: 0.50
# NETHER_GOLD_ORE: 1.25
ANCIENT_DEBRIS: 250.00
# HEART_OF_THE_SEA: 2500.00
# ENCHANTED_GOLDEN_APPLE: 25000.00
```
